### PR TITLE
Added Support in download test agent to specify multiple source and d…

### PR DESCRIPTION
**Pull request by acesiddhu**
Changes for patching the msi on test agent machine. Multiple source and destination paths

### CODE REVIEW

File: Tasks/DeployVisualStudioTestAgent/DownloadTestAgent.ps1 [hunk 1]
```diff
@@ -15,35 +15,44 @@ function ValidateSourceFile([string] $sourcePath)
...
+    if($isUri)
     {
-         throw "Robocopy failed to copy fail from $sourceDirectory to $destinationDirectory with a exit code $robocopyExitCode."
+        # Create the parent directory if it does not exist
+        $destinationDirectory = Split-Path -Path $destinationFile[$counter] -Parent
+        $isPresent = Test-Path $destinationDirectory
+        if(!$isPresent)
+        {
+            New-Item -ItemType Directory -Path $destinationDirectory
+        }
```
Comment (ajay-arora) [inline Tasks/DeployVisualStudioTestAgent/DownloadTestAgent.ps1]: another option is to use New-Item -Force option

File: Tasks/DeployVisualStudioTestAgent/DownloadTestAgent.ps1 [hunk 2]
```diff
@@ -15,35 +15,44 @@ function ValidateSourceFile([string] $sourcePath)
...
+
+        Write-Verbose -Message "Downloading test agent from $sourcePath to test machine."
```
Comment (ajay-arora) [inline Tasks/DeployVisualStudioTestAgent/DownloadTestAgent.ps1]: does source path variable get expanded like this ? we usually use -f option to format the variables.

File: Tasks/DeployVisualStudioTestAgent/DownloadTestAgent.ps1 [hunk 3]
```diff
@@ -15,35 +15,44 @@ function ValidateSourceFile([string] $sourcePath)
...
-$destinationDirectory = Split-Path -Path $destinationPath -Parent
-$isPresent = Test-Path $destinationDirectory
-if(!$isPresent)
-{
-    New-Item -ItemType Directory -Path $destinationDirectory
-}
-
-# Check if the given path is a valid Uri
-$isUri = [System.Uri]::IsWellFormedUriString($sourcePath, [System.UriKind]::Absolute)
+$source = $sourcePath.Split(";")
```
Comment (ajay-arora) [inline Tasks/DeployVisualStudioTestAgent/DownloadTestAgent.ps1]: pls update the help on the top of the file ...

File: Tasks/DeployVisualStudioTestAgent/DownloadTestAgent.ps1 [hunk 4]
```diff
@@ -15,35 +15,44 @@ function ValidateSourceFile([string] $sourcePath)
...
+$counter = 0;
+$destinationFile = $destinationPath.Split(";")
```
Comment (ajay-arora) [inline Tasks/DeployVisualStudioTestAgent/DownloadTestAgent.ps1]: can we test negative scenarios and see if the error is clear ?
for ex. split returning empty string, invalid syntax etc ?

**Comment by ajay-arora** [comment]
pls mention testing done

File: Tasks/DeployVisualStudioTestAgent/DownloadTestAgent.ps1 (hunk 1 above)
Comment (acesiddhu) [inline Tasks/DeployVisualStudioTestAgent/DownloadTestAgent.ps1]: not changing existing code

File: Tasks/DeployVisualStudioTestAgent/DownloadTestAgent.ps1 (hunk 3 above)
Comment (acesiddhu) [inline Tasks/DeployVisualStudioTestAgent/DownloadTestAgent.ps1]: done

File: Tasks/DeployVisualStudioTestAgent/DownloadTestAgent.ps1 (hunk 2 above)
Comment (acesiddhu) [inline Tasks/DeployVisualStudioTestAgent/DownloadTestAgent.ps1]: this works. -f is needed when you are formatting string  like {0}

**Comment by acesiddhu** [comment]
testing done if what I mentioned on vso pr.

File: Tasks/DeployVisualStudioTestAgent/DownloadTestAgent.ps1 [hunk 5]
```diff
@@ -15,35 +15,44 @@ function ValidateSourceFile([string] $sourcePath)
...
+        $counter++
+    }
+    else
+    {
+        ValidateSourceFile($sourcePath)
+        $sourceDirectory = Split-Path -Path $sourcePath -Parent
+        $sourceFileName = Split-Path -Path $sourcePath -Leaf
+
+        Write-Verbose -Message "Copying file from $sourcePath to test machine." -f $sourcePath
+	robocopy $sourceDirectory $destinationDirectory $sourceFileName /Z /mir /NP /Copy:DAT /R:10 /W:30
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/DownloadTestAgent.ps1]: would prefer if we printed out command lines we execute. helps user to debug

Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/DownloadTestAgent.ps1]: indentation is off

**Comment by allendm-msft** [comment]
signed off with comments

**Comment by acesiddhu** [comment]
tested all scenarios
RM, with and without custom path. directly downloading etc. found two bugs
1) we weren't resetting the event on powershell cmd side while downloading multiple exec. fixed that
2) made msi install sync call. 

no bugs in this script as such. Going ahead with checkin

**Pull request merged**